packages/shortest/src/core/runner/test-run-repository.ts

```diff
@@ -0,0 +1,411 @@
+  private static repositoriesByIdentifier = new Map<
+    string,
+    TestRunRepository
+  >();
+
+  public static getRepositoryForTestCase(
+    testCase: TestCase,
+  ): TestRunRepository {
+    const key = testCase.identifier;
+    if (!TestRunRepository.repositoriesByIdentifier.has(key)) {
+      const repository = new TestRunRepository(testCase);
+      TestRunRepository.repositoriesByIdentifier.set(key, repository);
+    }
+    return TestRunRepository.repositoriesByIdentifier.get(key)!;
+  }
+
+  private readonly testCase: TestCase;
+  private readonly lockFileName: string;
+  private readonly globalCacheDir: string;
+  private get lockFilePath(): string {
+    return path.join(this.globalCacheDir, this.lockFileName);
+  }
+  private readonly log: Log;
+  private readonly MAX_LOCK_ATTEMPTS = 10;
+  private readonly BASE_LOCK_DELAY_MS = 10;
+  private lockAcquired = false;
+  private testRuns: TestRun[] | null = null;
+
+  /**
+   * Creates a new test cache instance
+   * @param {TestCase} test - Test case to cache results for
+   */
+  constructor(testCase: TestCase, cacheDir = CACHE_DIR_PATH) {
+    this.log = getLogger();
+    this.log.trace("Initializing TestRunRepository", {
+      identifier: testCase.identifier,
+    });
+    this.testCase = testCase;
+    this.globalCacheDir = cacheDir;
+    this.lockFileName = `${this.testCase.identifier}.lock`;
+
+    // Register shared handlers and track this instance
+    TestRunRepository.activeRepositories =
+      registerSharedProcessHandlers(this.log) ||
+      TestRunRepository.activeRepositories;
+    TestRunRepository.activeRepositories.add(this);
+  }
+
+  async getRuns(): Promise<TestRun[]> {
+    if (this.testRuns !== null) {
+      return this.testRuns;
+    }
+
+    this.log.trace("Getting test runs", {
+      identifier: this.testCase.identifier,
+    });
+
+    const files = (await fs.readdir(this.globalCacheDir)).filter((f) =>
+      f.endsWith(`${this.testCase.identifier}.json`),
+    );
+
+    const loadedTestRuns: TestRun[] = [];
+
+    for (const file of files) {
+      try {
+        const filePath = path.join(this.globalCacheDir, file);
+        const content = await fs.readFile(filePath, "utf-8");
+        const cacheEntry = JSON.parse(content) as CacheEntry;
+        loadedTestRuns.push(
+          TestRun.fromCacheFile(this.testCase, file, cacheEntry),
+        );
+      } catch (error) {
+        this.log.error("Failed to load test run from cache", {
+          file,
+          ...getErrorDetails(error),
+        });
+      }
+    }
+    this.testRuns = loadedTestRuns;
+    return this.testRuns;
+  }
+
+  private resetTestRuns() {
+    this.testRuns = null;
+  }
+
+  async getLatestPassedRun(): Promise<TestRun | null> {
+    const testRuns = await this.getRuns();
+    const validTestRuns = testRuns.filter(
+      (testRun) =>
+        testRun.status === "passed" &&
+        testRun.version === TestRunRepository.VERSION &&
+        !testRun.fromCache,
```

Comment:
Test runs from cache don't have AI steps, so they are not useful for executing future runs off them.